power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/y_bus.hpp

```diff
@@ -331,22 +343,92 @@ template <bool sym> class YBus {
         for (Idx entry = 0; entry != nnz(); ++entry) {
             // start admittance accumulation with zero
             ComplexTensor<sym> entry_admittance{0.0};
+            IdxVector entry_param_shunt;
+            IdxVector entry_param_branch;
+            // loop over all entries of this position
+            for (Idx element = y_bus_entry_indptr[entry]; element != y_bus_entry_indptr[entry + 1]; ++element) {
+                auto param_idx = y_bus_element[element].idx;
+                if (y_bus_element[element].element_type == YBusElementType::shunt) {
+                    entry_admittance += shunt_param[param_idx];
+                    entry_param_shunt.push_back(param_idx);
+                } else {
+                    entry_admittance +=
+                        branch_param[param_idx].value[static_cast<Idx>(y_bus_element[element].element_type)];
+                    entry_param_branch.push_back(param_idx);
+                }
+            }
+            // assign
+            admittance[entry] = entry_admittance;
+            map_admittance_param_branch_.push_back(entry_param_branch);
+            map_admittance_param_shunt_.push_back(entry_param_shunt);
+        }
+        // move to shared ownership
+        admittance_ = std::make_shared<ComplexTensorVector<sym>>(std::move(admittance));
+
+        parameters_changed(true);
+    }
+
+    IdxVector increments_to_entries(auto const& math_model_param_incrmt) {
+        // construct affected entries
+        IdxVector affected_entries;
+
+        auto query_params_in_map = [&affected_entries](auto const& params_to_change, auto const& mapping) {
+            for (size_t i = 0; i < mapping.size(); ++i) {
+                if (std::ranges::any_of(mapping[i], [&](Idx val) {
+                        return std::ranges::find(params_to_change, val) != params_to_change.end();
+                    })) {
+                    affected_entries.push_back(static_cast<int>(i));
+                }
+            }
+        };
+
+        query_params_in_map(math_model_param_incrmt.branch_param_to_change, map_admittance_param_branch_);
+        query_params_in_map(math_model_param_incrmt.shunt_param_to_change, map_admittance_param_shunt_);
+        return affected_entries;
+    }
+
+    /**
+     * @brief Updates the admittance of the y_bus according to what's changed in math_model.
+     *
+     * @param math_model_param Shared pointer to the constant math_model parameters.
+     * @param math_model_param_incrmt Shared pointer to the constant mathematical model parameters .
+     */
+    void update_admittance_increment(std::shared_ptr<MathModelParam<sym> const> const& math_model_param,
+                                     MathModelParamIncrement const& math_model_param_incrmt) {
+        // swap the old cached parameters
+        math_model_param_ = math_model_param;
+
+        if (admittance_.use_count() != 1) {
```

Comment:
right... ok that's a good one.

how bad is it to copy once for all instances? from the original implementation, we already used to do that (by recreating)... Then are no shared matrix elements and, instead, only the ybus structure is shared. would make the logic easier.

Reply:
So what are your thoughts @TonyXiang8787 @mgovers , should we include race condition prevention to this PR as well?

Reply:
I think to accommodate that change, some logic across threads need to be slightly modified. They we wouldn't be needing the signaling mechanism, would we? @mgovers

Reply:
the signaling mechanism is unrelated. the signaling is necessary because a (specifically the iterative current) solver needs to know whether some smart tricks (e.g. pre-caching) needs to happen again - which is necessary in case the ybus changes.

it is not possible to make the solver ask to ybus "did you change", because multiple calculation method solvers might be using the same ybus, and if one of them asks "did you change", the other cannot ask again